Newmark

oriolcg opened the pull request:
This is a first implementation of second-order ODEs with the Newmark-beta solver. I followed the approach of defining new structures to handle the needs for 2nd-order operators, instead of trying to extend what we had for first order ODEs.

The basic idea is that we need to store additional temporary variables for the first and second time derivative (denoted as `v` and `a` as in velocity and acceleration). 

@santiagobadia @fverdugo , please take a look at this PR and let me know if this approach looks good to you.

santiagobadia commented on src/ODETools/Newmark.jl:
```diff
@@ -0,0 +1,95 @@
+  v1 = op.γ/(op.β*op.dt)*(u1-op.u0) + (1-op.γ/op.β)*op.v0 + op.dt*(1-op.γ/(2*op.β))*op.a0
+  z = zero(eltype(A))
+  fill_entries!(A,z)
+  jacobian_and_jacobian_t!(A,op.odeop,op.t1,u1,v1,a1,op.γ/(op.β*op.dt),1.0/(op.β*op.dt^2),cache)
```
Probably, this name is misleading now, since we have jacobian_tt, haven't we?

oriolcg replied:
Changed the name to `jacobians` in the new approach.

santiagobadia commented on src/ODETools/ODEOperators.jl:
```diff
@@ -8,13 +8,15 @@ struct Nonlinear <: OperatorType end
 """
 abstract type ODEOperator{C<:OperatorType} <: GridapType end
+abstract type SecondOrderODEOperator{C} <: ODEOperator{C} end
 
```
Now, the documentation is not updated, `A(t,u,∂tu,∂ttu,...,∂t^nu). If it is not much work, we could consider the order a template parameter.

oriolcg replied:
I updated the documentation with the new approach, which is general for N-th order ODEs.

santiagobadia commented on src/ODETools/ODESolutions.jl:
```diff
@@ -70,3 +70,55 @@ function Base.iterate(sol::GenericODESolution, state)
+  a0::AbstractVector
+  t0::Real
+  tF::Real
+end
```
As above, GenericODESolution could have the order N as template parameter and a tuple of N abstract vectors. This way, I think you don't need to replicate the code for N=1 and N=2. On the other side, I don't think we want to go more than second order, so it is probably too much.

oriolcg replied:
Done in the new approach

santiagobadia commented on src/ODETools/ODESolvers.jl:
```diff
@@ -23,6 +23,30 @@ function solve(
+  t0::Real) # -> (uF,tF,cache)
+  solve_step!(uF,solver,op,u0,v0,a0,t0,nothing)
+end
+
```
Why `a0` is needed here? In general, you only need first order derivatives as initial value of a second order operator. Besides, it is not consistent with the first order case.

If we can eliminate `a0`, again, working with a template of N-vectors, I think we would not need to replicate code, one implementation would be enough.

oriolcg replied:
The general Newmark-beta method needs `u0`, `v0` and `a0`. It is also consistent with other multi-step methods for first order ODEs like the generalized-alpha, which needs `u0` and `v0`, or BDF2 ...

The new implementation can handle an arbitrary number of initial values.

santiagobadia commented on src/TransientFETools/ODEOperatorInterfaces.jl:
```diff
@@ -64,3 +64,125 @@ function jacobian_and_jacobian_t!(J::AbstractMatrix,op::ODEOpFromFEOp,t::Real,uh
+  uht = EvaluationFunction(Uht,uhtF)
+  uhtt = EvaluationFunction(Uhtt,uhttF)
+  jacobian_and_jacobian_t!(J,op.feop,t,uh,uht,uhtt,dut_u,dutt_u,ode_cache)
+end
```
I would probably put it into another file. Again, I think that with the changes I propose above there is no need to replicate all the code from N=1 to N=2.

oriolcg replied:
No need for new functions. The functions have been generalized to N-th order ODEs

santiagobadia commented on src/TransientFETools/Transient2ndOrderFEOperators.jl:
```diff
@@ -0,0 +1,188 @@
+  trial_t
+  trial_tt
+  test::FESpace
+end
```
We could consider here a tuple of Jacobians (N+1) and no need to replicate. Idem with trials.

oriolcg replied:
Done

santiagobadia commented on src/TransientFETools/TransientFEOperators.jl:
```diff
@@ -156,7 +156,9 @@ function residual!(b::AbstractVector,op::TransientFEOperatorFromWeakForm,
 function allocate_jacobian(op::TransientFEOperatorFromWeakForm,uh::FEFunction,cache)
-  matdata = matdata_jacobian(op,0.0,uh,uh)
+  matdata_j = matdata_jacobian(op,0.0,uh,uh)
+  matdata_jt = matdata_jacobian_t(op,0.0,uh,uh,0.0)
```
Not Jtt?

oriolcg replied:
This was a fix for 1st order ODEs, for 2nd order ODEs the operator type was `Transient2ndOrderFEOperatorFromWeakForm` and had an alternative implementation. The new function now works for N-th order.

santiagobadia commented on src/TransientFETools/TransientFESolutions.jl:
```diff
@@ -22,6 +22,20 @@ function TransientFESolution(
+  trial = get_trial(op)
+
+  TransientFESolution(ode_sol, trial)
+end
```
Idem as before, usign a N+1-tuple no need to replicate

oriolcg replied:
Done

santiagobadia commented on src/TransientFETools/TransientFESolvers.jl:
```diff
@@ -10,6 +10,11 @@ function solve(
+function solve(
+  solver::TransientFESolver,op::TransientFEOperator,u0,v0,a0,t0::Real,tf::Real)
+  TransientFESolution(solver,op,u0,v0,a0,t0,tf)
+end
```
Idem

oriolcg replied:
Done

santiagobadia commented on test/TransientFEsTests/Transient2ndOrderFEOperatorsTests.jl:
```diff
@@ -0,0 +1,75 @@
+  @test el2 < tol
+end
+
+end
```
I would add a problem that can be computed exactly to check the implementation

oriolcg replied:
This should be the case. Here it's solving a quadratic function in space and time with quadratic shape functions and 2nd order Newmark-beta scheme.

oriolcg replied:
That should be the case. Here it's solving a quadratic function in space and time with quadratic shape functions and a second order Newmark-beta scheme.

santiagobadia commented:
Thanks @oriolcg for your PR! Great job!

I have proposed some changes. We can discuss it. My main concern is whether we need to replicate code or we could just re-define the objects such that order is a template parameter and we have tuples of N+1 arrays, vectors, trial spaces, etc. It seems quite easy at least conceptually.

Anyway, if you consider it is too much work, we can keep this implementation. We probably agree that if we want higher-order ODEs, this approach is not appropriate, but we won't need them anyway.

There are some other minor comments.

fverdugo commented on src/ODETools/ODEOperators.jl:
```diff
@@ -8,13 +8,15 @@ struct Nonlinear <: OperatorType end
 or constant in time.
 """
 abstract type ODEOperator{C<:OperatorType} <: GridapType end
+abstract type SecondOrderODEOperator{C} <: ODEOperator{C} end
```
Just a minor thing: I would not extend `SecondOrderODEOperator` from `ODEOperator` since they have different interfaces. I.e., the code implemented for `ODEOperator` will not work if you pass a `SecondOrderODEOperator`.

oriolcg replied:
No need for another type in the new implementation

fverdugo reviewed (commented):
Thanks for the PR @oriolcg !

oriolcg commented:
Hi @santiagobadia @fverdugo, I modified the approach and refactored the code in such a way that it can handle N-th order ODEs. Basically, I added tuples where needed and generalized as much as possible without compromising too much the current APIs.

The only thing that would need to be modified at the driver level is the way of defining the residuals and jacobians. Instead of `res(t,u,ut,v)` now is `res(t,(u,ut),v)`. This is needed to generalize the calls for an arbitrary number of time derivatives. This also opens the door to the API that Francesc proposed: `res(t,u,v) = m(∂tt(u),v) + c(∂t(u),v) + a(u,v) - b(v,t)`.

Please, take a look at the latest commit https://github.com/gridap/GridapODEs.jl/pull/51/commits/ea237001a439de07b30e697ab31bd9b218cd8ab0 and let me know if there are further changes that need to be done.

santiagobadia commented on src/ODETools/ODEOperators.jl:
```diff
@@ -8,76 +8,88 @@ struct Nonlinear <: OperatorType end
 """
-Add the contribution of both `jacobian!` and `jacobian_t!`,i.e.,
-[∂A/∂u](t,u,∂tu)+γ*[∂A/∂(∂tu)](u,∂tu)
+Add the contribution of both all jacobians ,i.e., ∑ᵢ γ_i*[∂A/∂(∂t^iu)](t,u,∂tu,...,,∂t^Nu)
```
both out

oriolcg replied:
done in https://github.com/gridap/GridapODEs.jl/commit/8e371a3e8be38ed1b2875c75e873e2c0db7c62de

santiagobadia commented on src/TransientFETools/ODEOperatorInterfaces.jl:
```diff
@@ -7,60 +7,90 @@ struct ODEOpFromFEOp{C} <: ODEOperator{C}
+  for i in 1:get_order(op)+1
+    xh = (xh...,EvaluationFunction(Xh[i],xhF[i]))
+  end
+  jacobian!(A,op.feop,t,xh,i,γᵢ,ode_cache)
```
I see that now we include the integer i to pick the Jacobian we want and a real number (also for i=1). We could probably put some documentation on the meaning of these things in case not already been done. It is a little bit messy to do the PR here with so many changes.

oriolcg replied:
It was done at the abstract method. I improved the documentation and added it to the ODEOperatorInterfaces.

santiagobadia commented:
Hi @oriolcg a couple of very minor comments. We could also add an issue for creating the API @fverdugo suggested.

santiagobadia commented:
No, we can keep it for later. I accept the PR. Thanks!
